pkg/server/plugin/upstreamauthority/spire/spire.go

```diff
@@ -70,59 +92,166 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
 	return &plugin.GetPluginInfoResponse{}, nil
 }
 
-func (m *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
-	ctx := stream.Context()
+func (m *Plugin) SetLogger(log hclog.Logger) {
+	m.log = log
+}
 
-	wCert, wKey, wBundle, err := m.getWorkloadSVID(ctx, m.config)
+func (m *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
+	err := m.subscribeToPolling(stream.Context())
 	if err != nil {
 		return err
 	}
+	defer m.unsubscribeToPolling()
 
-	conn, err := m.newNodeClientConn(ctx, wCert, wKey, wBundle)
+	certChain, err := m.submitCSRUpstreamCA(stream.Context(), request.Csr)
 	if err != nil {
 		return err
 	}
-	nodeClient := node.NewNodeClient(conn)
-	defer conn.Close()
 
-	certChain, bundle, err := m.submitCSRUpstreamCA(ctx, nodeClient, request.Csr)
-	if err != nil {
-		return err
+	rootCAs := []*common.Certificate{}
+	rawChain := certsToRawCerts(certChain)
+	ticker := time.NewTicker(internalPollFreq)
+	defer ticker.Stop()
+	for {
+		newRootCAs := m.getBundle().RootCas
+		if !areRootsEqual(rootCAs, newRootCAs) {
+			rootCAs = newRootCAs
+			err := stream.Send(&upstreamauthority.MintX509CAResponse{
+				X509CaChain:       rawChain,
+				UpstreamX509Roots: commonCertsToRawCerts(rootCAs),
+			})
+			if err != nil {
+				m.log.Error("Cannot send X.509 CA chain and roots", "error", err)
+				return err
+			}
+			if rawChain != nil {
+				rawChain = nil
+			}
+		}
+		select {
+		case <-ticker.C:
+		case <-stream.Context().Done():
+			return nil
+		}
 	}
-
-	return stream.Send(&upstreamauthority.MintX509CAResponse{
-		X509CaChain:       certsToRawCerts(certChain),
-		UpstreamX509Roots: certsToRawCerts(bundle.RootCAs()),
-	})
 }
 
 func (m *Plugin) PublishJWTKey(req *upstreamauthority.PublishJWTKeyRequest, stream upstreamauthority.UpstreamAuthority_PublishJWTKeyServer) error {
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
-	ctx := stream.Context()
-
-	wCert, wKey, wBundle, err := m.getWorkloadSVID(ctx, m.config)
+	err := m.subscribeToPolling(stream.Context())
 	if err != nil {
 		return err
 	}
+	defer m.unsubscribeToPolling()
 
-	conn, err := m.newNodeClientConn(ctx, wCert, wKey, wBundle)
+	err = m.pushAndSetInitialKeys(stream.Context(), req.JwtKey)
 	if err != nil {
 		return err
 	}
-	nodeClient := node.NewNodeClient(conn)
-	defer conn.Close()
 
-	resp, err := nodeClient.PushJWTKeyUpstream(ctx, &node.PushJWTKeyUpstreamRequest{JwtKey: req.JwtKey})
-	if err != nil {
-		return err
+	keys := []*common.PublicKey{}
+	ticker := time.NewTicker(internalPollFreq)
```

Comment:
nit: can the speed of the unit tests be improved by having these tickers use the clock as well?

Reply:
I'm ok with this being addressed in a follow up PR. We'd like this in for 0.10.0 and that release timeline trumps a few additional seconds in a unit test.

Reply:
Perfect, I'll let that improvement for another PR